### PR TITLE
To be released by 2/23

### PR DESCRIPTION
**Pull request by nibinqtl**
@burtonquantil and @marcfishelquantil , please review the new file added by me: REST API acceleration
Bin, I (Marc) reviewed and replied.

### CODE REVIEW

File: docs/recipes/api-acceleration.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+## REST API Acceleration
+
+API calls are usually considered dynamic HTTP requests, since the responses are generated
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:3]: Remove the comma after "API calls are usually considered dynamic HTTP requests".

File: docs/recipes/api-acceleration.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+by the server in real time based on some input parameters supplied in the request. As we
+mentioned in the [FAQ](/docs/edge-logic/faq#what-about-dynamic-content), dynamic requests
+such as REST API calls can be very effectively accelerated by CDN360. For example, the user
+of a mobile or web app may reload the same page repeatedly, which may generate lots of
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:7]: Suggestion:  "For example, the user of a mobile or web app may repeatedly reload the same page generating lots of "
What do you think @marcfishelquantil  ?

File: docs/recipes/api-acceleration.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
+## REST API Acceleration
+
+API calls are usually considered dynamic HTTP requests, since the responses are generated
+by the server in real time based on some input parameters supplied in the request. As we
+mentioned in the [FAQ](/docs/edge-logic/faq#what-about-dynamic-content), dynamic requests
+such as REST API calls can be very effectively accelerated by CDN360. For example, the user
+of a mobile or web app may reload the same page repeatedly, which may generate lots of 
+duplicated API calls in a short period of time. Some caching using CDN can result in a
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:8]: Replace "duplicated" with "duplicate".

Comment (burtonquantil) [reply docs/recipes/api-acceleration.md:8]: Suggestion: Some caching using **a** CDN  can drastically improve the performance.

File: docs/recipes/api-acceleration.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+drastic improvement of the performance. In this article, we are going to use an example to illustrate how
+to use the portal to create a property to accelerate an API server.
+
+Here are the assumptions about this task:
+* The API server to be accelerated has a hostname: `api.company.com`. The server requires
+the request `Host` header to carry this value.
+* You currently setup the DNS server to resolve this hostname to 2 IP addresses: `1.1.1.1`
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:15]: Change "setup" to the verb "set up".

I prefer to see "two" versus "2", but that's minor.

File: docs/recipes/api-acceleration.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+* The client is using the `Authorization` request header to pass the credential to the
+server following the format of [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication).
+This makes it easy for the CDN360 servers to obtain the API user name with the built-in
+variable `$remote_user`. Please notice that this does not mean you have to use the basic
+authentication which transfers the secret password in clear text. You are free to use
+algorithms like [AWSv2](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html)
+to generate the signature after the colon.
+* All the input variables to the API server are specified in the request query string.
+
+In order to use CDN360 to accelerate this service, we need to do the following:
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:27]: Replace "In order to use" with "To use" .
Marc mentioned in another place that "In order" is really unnecessary verbiage.

File: docs/recipes/api-acceleration.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+* Create a new DNS record `api-origin.company.com` to point to the 2 IP addresses. The 
+CDN360 servers need to use it to reach the origin servers. The name `api.company.com` can
+no longer be used because we later will CNAME it to a CDN360 edge hostname to direct
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:30]: Change "later will" to "will later".

Is it proper to use "CNAME" as a verb?

File: docs/recipes/api-acceleration.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+client's traffic to the platform to be accelerated.
+* No API service should be running without the protection of TLS encryption. You need to
+upload the certificate for `api.company.com` to the CDN360 platform. We recommend the use
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:33]: Suggest replacing "the use of " to "using"

File: docs/recipes/api-acceleration.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+of [Let's Encrypt](/docs/portal/certificates/auto-renewal) to automatically renew the certificate.
+<p align=center><img src="/docs/resources/images/recipes/api/upload-certificate.png" alt="upload certificate" width="700"></p>
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:35]: in sprint 101 which we will release soon, we changed the certificate selection to radio buttons per your request. NCP-2164

File: docs/recipes/api-acceleration.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+* Go to the CDN360 portal to create an property to accelerate this API service. The
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:37]: Change "an property" to "a property".

File: docs/recipes/api-acceleration.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+of [Let's Encrypt](/docs/portal/certificates/auto-renewal) to automatically renew the certificate.
+<p align=center><img src="/docs/resources/images/recipes/api/upload-certificate.png" alt="upload certificate" width="700"></p>
+
+* Go to the CDN360 portal to create an property to accelerate this API service. The 
+important thing in this step is to enter the correct hostname to be accelerated: `api.company.com`.
+<p align=center><img src="/docs/resources/images/recipes/api/create-property.png" alt="create property" width="720"></p>
+
+* Enter the information about the origin. A few things to note on this page: The server
+is specified with the new DNS record we just created. We are enforcing the HTTPS protocol
+to reach the origin to ensure security. The `Host` header value is specified to be the
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:43]: I suggest simplifying  "is specified to be " to simply "is ".

File: docs/recipes/api-acceleration.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+one required by the origin. In this case, we can actually leave it empty because by
+default, CDN360 will pass the `Host` header value received from client to the origin. We also
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:45]: I suggest changing   "from client" to "from the client"

File: docs/edge-logic/faq.md [hunk 12]
```diff
@@ -31,7 +31,7 @@ If you need to add, modify, or delete some header fields in the response to clie
 ```nginx
 add_header CDN-Name Quantil;
 ```
-We also created a proprietary directive [`origin_header_modify`](</docs/edge-logic/supported-directives.md#origin_header_modify>) to manipulate the response header from origin, before any other processing of the response. This can be very useful if you need to override some header value (such as cache time) from the origin that may affect the CDN servers' behavior.
+We also created a proprietary directive [`origin_header_modify`](</docs/edge-logic/supported-directives.md#origin_header_modify>) to manipulate the response header from origin before any other processing of the response. This can be very useful if you need to override some header value (such as cache time) from the origin that may affect the CDN servers' behavior.
```
Comment (marcfishelquantil) [inline docs/edge-logic/faq.md:35]: Please change line 34 as follows (changes are in CAPS): "We also created a proprietary directive [`origin_header_modify`](</docs/edge-logic/supported-directives.md#origin_header_modify>) to manipulate the response header from THE origin PRIOR TO processing the response. This can be very useful if you need to override A header value (such as cache time) from the origin that may affect the CDN servers' behavior."

File: docs/recipes/api-acceleration.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+chose "Always Direct" to reach the origin without going through a parent cache. This is because we want to
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:46]: Suggestion: combine sentences.    
"... through a parent cache because we want to minimize latency and know there will not be a cache hit across different servers".

File: docs/recipes/api-acceleration.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+minimize latency and we know there is not going to be cache hit across different servers.
+<p align=center><img src="/docs/resources/images/recipes/api/origin.png" alt="create origin" width="600"></p>
+
+* Enter the following code into the Edge Logic field. The important thing to note here is
+that the API user name and client IP are added to the cache key. This ensures that the 
+cached content will be only served to the same user from the same IP address. Combined
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:52]: I think "will be only served" should be "will be served only" ?  @marcfishelquantil

File: docs/portal/accessing-portal/navigating-ui.md [hunk 15]
```diff
@@ -28,6 +28,7 @@ The [Reports page](</docs/portal/reports.md>) allows you to generate reports abo
 - Traffic volume and bandwidth to your hostnames.
 - Number of requests made to the files of your property.
 - Percentage of each HTTP status code returned as a result of requests to your content.
+- Content that has been requested and when, the IP address from which requests were made, the value of the User-Agent HTTP request header, and the size of the content requested.
```
Comment (marcfishelquantil) [inline docs/portal/accessing-portal/navigating-ui.md:31]: Looks fine to me. A more verbose definition could be: "- Content that has been requested and THE START AND END TIME OF THE REQUEST, the IP address from which requests were made, the value of the User-Agent HTTP request header, and the size of the content requested."

or

"Content that has been requested and THE TIME when IT WAS REQUESTED, the IP address from which requests were made, the value of the User-Agent HTTP request header, and the size of the content requested."

File: docs/recipes/api-acceleration.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+with HTTPS and the short cache time of 1 minute, this should be safe enough for the vast
+majority of applications in the industry. Another thing to notice is that we allow the
+clients to use the `Cache-Control: no-cache` header field to by pass cache. When a CDN360
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:55]: change "by pass cache" to "bypass the cache"

File: docs/recipes/api-acceleration.md [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+* Make sure the correct certificate is attached to this property. You can configure the
+TLS min/max versions and cipher suites based on your security requirements:
+<p align=center><img src="/docs/resources/images/recipes/api/attach-certificate.png" alt="attach certificate" width="600"></p>
+* Optionally test the above property [in staging](/docs/portal/edge-configurations/testing-property#testing-property-in-staging)
+and then deploy to production.
+* If you still don't have one, create an "edge hostname" based on your intended delivery
+destination and performance/cost requirement. Then, update the DNS record of `api.company.com`
+to CNAME to this edge hostname.
+
+Any request to `api.company.com` will now be routed to the CDN360 platform. Every cache-miss
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:89]: I think "cache-miss"  should be "cache miss".

File: docs/recipes/api-acceleration.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+or expiration will be forwarded to the origin to validate the credential and fetch the latest content.
+If the same request comes again from the same user and the same IP address before the 
+cached copy expires, the content will be served immediately by the CDN360 server, with
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:92]: remove the comma after "CDN360 server" .
Change "turn-around" to "turnaround"

File: docs/recipes/api-acceleration.md [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+much smaller turn-around time. The end user will experience an improved performance and
```
Comment (burtonquantil) [inline docs/recipes/api-acceleration.md:93]: Change "an improved performance" to "improved performance"

**Review by burtonquantil — CHANGES_REQUESTED** [review]
I made some suggestions.

File: docs/portal/identities-and-access/overview.md [hunk 20]
```diff
@@ -36,7 +36,7 @@ The following figure shows the key elements on the page, and the table following
 
 | **Fields** | **Description** |
 | :----------: | --------------- |
-| 1 | Identity and access management tasks are organized into four tabs:<ul> <li>**Users** allows administrators to manage users. Operators and viewers are limited to seeing other portal users. This tab appears when you click **Identity & Access Management**. <li>**Roles** allows administrators to manage roles. Operators and viewers are limited to seeing roles. <li>**Audit Logs** allows administrators to manage audit logs. This tab is not available to operators and viewers. <li>**Security**   allows administrators to define security for the portal. This tab is not available to operators and viewers.|
+| 1 | Identity and access management tasks are organized into the following tabs:<ul> <li>**Users** allows administrators to manage users. Operators and viewers are limited to seeing other portal users. This tab appears when you click **Identity & Access Management**. <li>**Roles** allows administrators to manage roles. Operators and viewers are limited to seeing roles.|
```
Comment (marcfishelquantil) [inline docs/portal/identities-and-access/overview.md:39]: Looks fine to me.

File: docs/portal/reports.md [hunk 21]
```diff
@@ -28,6 +28,7 @@ The following report types are supported:
 | Traffic Volume Summary| Summarizes the edge traffic volume.                               |
 | CPU Time| Reports the CPU time consumed in each time interval to serve your properties. You can toggle the unit between **seconds** and **number of cores**. For example, if 3000 seconds of CPU time are reported in a 5-minute interval, it means your properties are maxing out an average of 10 (=3000s/5m) physical CPU cores during that period.                               |
 | Edge Hostname Request| Shows the number of requests made to your [edge hostnames](</docs/portal/traffic-management/creating-edge-hostname.md>) over a period of time.                               |
+| Access Log Download| Shows what content has been requested and when it was requested, the IP address from which requests were made, the value of the User-Agent HTTP request header, and the size of the content requested.                    |
```
Comment (marcfishelquantil) [inline docs/portal/reports.md:31]: Please see my comment above on line 31.

File: docs/portal/reports.md [hunk 22]
```diff
@@ -40,6 +41,7 @@ The following report types are supported:
 | --------------- | ----------------------------------------------------- |
 | Report Type     | Select the type of report you want to generate.       | 
 | Date Range      | Select the start and end dates and the time for the report. |
+| Hostname | For Access Log Download reports, select a property hostname. Logs of visitors to the hostname will be generated for you to view and download.
```
Comment (marcfishelquantil) [inline docs/portal/reports.md:44]: Looks fine to me.

File: docs/recipes/api-acceleration.md (hunk 2 above)
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:7]: Since you use "may" earlier in this sentence, how about rewording the text string from "..which may generate lost of..." to "...which can generate a lot of..."

File: docs/recipes/api-acceleration.md [hunk 23]
```diff
@@ -0,0 +1,95 @@
+## REST API Acceleration
+
+API calls are usually considered dynamic HTTP requests, since the responses are generated
+by the server in real time based on some input parameters supplied in the request. As we
+mentioned in the [FAQ](/docs/edge-logic/faq#what-about-dynamic-content), dynamic requests
+such as REST API calls can be very effectively accelerated by CDN360. For example, the user
+of a mobile or web app may reload the same page repeatedly, which may generate lots of 
+duplicated API calls in a short period of time. Some caching using CDN can result in a
+drastic improvement of the performance. In this article, we are going to use an example to illustrate how
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:9]: Please change "drastic improvement of the performance" to "significant performance improvement" or "significant improvement in performance."

File: docs/recipes/api-acceleration.md (hunk 4 above)
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:15]: setup should be two words (set up).

File: docs/recipes/api-acceleration.md [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+to use the portal to create a property to accelerate an API server.
+
+Here are the assumptions about this task:
+* The API server to be accelerated has a hostname: `api.company.com`. The server requires
+the request `Host` header to carry this value.
+* You currently setup the DNS server to resolve this hostname to 2 IP addresses: `1.1.1.1`
+and `1.1.1.2`
+* The API server is using the standard HTTP methods: GET, POST, PUT, DELETE, PATCH
+* The client is using the `Authorization` request header to pass the credential to the
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:18]: In lines 17 & 18, please consider using "uses" instead of "is using." Please add a period after "PATCH" at the end of the sentence on line 17.

File: docs/recipes/api-acceleration.md [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+server following the format of [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication).
+This makes it easy for the CDN360 servers to obtain the API user name with the built-in
+variable `$remote_user`. Please notice that this does not mean you have to use the basic
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:21]: The words "Please notice that" are extraneous and can be removed.

File: docs/recipes/api-acceleration.md (hunk 5 above)
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:27]: Please remove "In order." Also, please remove "we need to."

File: docs/recipes/api-acceleration.md [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+authentication which transfers the secret password in clear text. You are free to use
+algorithms like [AWSv2](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html)
+to generate the signature after the colon.
+* All the input variables to the API server are specified in the request query string.
+
+In order to use CDN360 to accelerate this service, we need to do the following:
+* Create a new DNS record `api-origin.company.com` to point to the 2 IP addresses. The 
+CDN360 servers need to use it to reach the origin servers. The name `api.company.com` can
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:29]: Please change "need to use it" to "use the DNS record"

File: docs/recipes/api-acceleration.md [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+no longer be used because we later will CNAME it to a CDN360 edge hostname to direct
+client's traffic to the platform to be accelerated.
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:31]: please change " later will CNAME it to a" to " CNAME it later to a"

File: docs/recipes/api-acceleration.md (hunk 7 above)
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:33]: Please change "the use of" to "you use"

File: docs/recipes/api-acceleration.md [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+* No API service should be running without the protection of TLS encryption. You need to
+upload the certificate for `api.company.com` to the CDN360 platform. We recommend the use
+of [Let's Encrypt](/docs/portal/certificates/auto-renewal) to automatically renew the certificate.
+<p align=center><img src="/docs/resources/images/recipes/api/upload-certificate.png" alt="upload certificate" width="700"></p>
+
+* Go to the CDN360 portal to create an property to accelerate this API service. The 
+important thing in this step is to enter the correct hostname to be accelerated: `api.company.com`.
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:38]: Consider changing "The important thing in this step is..." to "It is important to" or "Be sure to".

File: docs/recipes/api-acceleration.md [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+<p align=center><img src="/docs/resources/images/recipes/api/create-property.png" alt="create property" width="720"></p>
+
+* Enter the information about the origin. A few things to note on this page: The server
+is specified with the new DNS record we just created. We are enforcing the HTTPS protocol
+to reach the origin to ensure security. The `Host` header value is specified to be the
+one required by the origin. In this case, we can actually leave it empty because by
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:44]: Please add a comma after "because".

File: docs/recipes/api-acceleration.md (hunk 14 above)
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:52]: Please change "will be only served to" to "will be served only to".

File: docs/recipes/api-acceleration.md [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+default, CDN360 will pass the `Host` header value received from client to the origin. We also
+chose "Always Direct" to reach the origin without going through a parent cache. This is because we want to
+minimize latency and we know there is not going to be cache hit across different servers.
+<p align=center><img src="/docs/resources/images/recipes/api/origin.png" alt="create origin" width="600"></p>
+
+* Enter the following code into the Edge Logic field. The important thing to note here is
+that the API user name and client IP are added to the cache key. This ensures that the 
+cached content will be only served to the same user from the same IP address. Combined
+with HTTPS and the short cache time of 1 minute, this should be safe enough for the vast
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:53]: Please consider changing "safe enough" to "sufficiently safe".

File: docs/recipes/api-acceleration.md [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+majority of applications in the industry. Another thing to notice is that we allow the
+clients to use the `Cache-Control: no-cache` header field to by pass cache. When a CDN360
+server sees this field value, it will go directly to origin without looking up the cache.
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:56]: Should be "to the origin".

File: docs/recipes/api-acceleration.md [hunk 32]
```diff
@@ -0,0 +1,95 @@
...
+Last but not least, we enabled the [Fast Route to origin](/docs/edge-logic/supported-directives#origin_fast_route)
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:57]: "Last but not least" is too colloquial IMHO. How about "Lastly" instead?

File: docs/recipes/api-acceleration.md [hunk 33]
```diff
@@ -0,0 +1,95 @@
...
+  proxy_cache_valid 1m; #200, 301, and 302 responses will be cached for 1m
+  origin_fast_route on; #enable the Fast Route to origin
+}
+```
+* Make sure the correct certificate is attached to this property. You can configure the
+TLS min/max versions and cipher suites based on your security requirements:
+<p align=center><img src="/docs/resources/images/recipes/api/attach-certificate.png" alt="attach certificate" width="600"></p>
+* Optionally test the above property [in staging](/docs/portal/edge-configurations/testing-property#testing-property-in-staging)
+and then deploy to production.
+* If you still don't have one, create an "edge hostname" based on your intended delivery
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:85]: Still don't have what? Please replace "one" with "edge hostname".

File: docs/recipes/api-acceleration.md [hunk 34]
```diff
@@ -0,0 +1,95 @@
...
+destination and performance/cost requirement. Then, update the DNS record of `api.company.com`
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:86]: Please remove the comma after "Then".

File: docs/recipes/api-acceleration.md (hunk 19 above)
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:93]: Please add "a" between "with" and "smaller" as follows: "with a much smaller". Or how about replacing "smaller" with "faster" as in: "with a much faster".

File: docs/recipes/api-acceleration.md [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+to CNAME to this edge hostname.
+
+Any request to `api.company.com` will now be routed to the CDN360 platform. Every cache-miss
+or expiration will be forwarded to the origin to validate the credential and fetch the latest content.
+If the same request comes again from the same user and the same IP address before the 
+cached copy expires, the content will be served immediately by the CDN360 server, with
+much smaller turn-around time. The end user will experience an improved performance and 
+the origin server will be alleviated from generating the same content repeatedly in a
```
Comment (marcfishelquantil) [inline docs/recipes/api-acceleration.md:94]: Instead of "be alleviated from generating" how about "not have to generate".

**Comment by nibinqtl** [comment]
I think I addressed all suggestions made by @burtonquantil and @marcfishelquantil . Please take another look.

**Review by burtonquantil — APPROVED** [review]
OK to merge.
